src/main/java/org/folio/service/budget/BudgetService.java

```diff
@@ -237,7 +175,91 @@ public Future<Void> closeBudgets(LedgerFiscalYearRollover rollover, DBConn conn)
 
-    return budgetDAO.updateBatchBudgets(sql, conn)
+    return budgetDAO.updateBatchBudgetsBySql(sql, conn)
+      .mapEmpty();
+  }
+
+  public Future<List<Budget>> getBudgetsByFiscalYearIdsAndFundIdsForUpdate(
+      Map<String, Set<String>> fiscalYearIdToFundIds, DBConn conn) {
+    StringBuilder sb = new StringBuilder();
+    sb.append("SELECT jsonb FROM ")
+      .append(getFullTableName(conn.getTenantId(), BUDGET_TABLE))
+      .append(" WHERE ");
+    Iterator<String> fiscalYearIterator = fiscalYearIdToFundIds.keySet().iterator();
+    while (fiscalYearIterator.hasNext()) {
+      String fiscalYearId = fiscalYearIterator.next();
+      Set<String> fundIds = fiscalYearIdToFundIds.get(fiscalYearId);
+      sb.append("(jsonb->>'fiscalYearId' = '")
+        .append(fiscalYearId)
+        .append("' AND (");
+      Iterator<String> fundIterator = fundIds.iterator();
+      while (fundIterator.hasNext()) {
+        String fundId = fundIterator.next();
+        sb.append("jsonb->>'fundId' = '")
+          .append(fundId)
+          .append("'");
+        if (fundIterator.hasNext()) {
+          sb.append(" OR ");
+        }
+      }
+      sb.append("))");
+      if (fiscalYearIterator.hasNext()) {
+        sb.append(" OR ");
+      }
+    }
+    sb.append(" FOR UPDATE");
+    String sql = sb.toString();
+    return budgetDAO.getBudgetsBySql(sql, Tuple.tuple(), conn);
+  }
+
+  private Future<Void> deleteAllocationTransactions(Budget budget, DBConn conn) {
+    logger.debug("deleteAllocationTransactions:: Trying to delete allocation transactions with fund id {}", budget.getFundId());
+
+    String sql ="DELETE FROM "+ getFullTableName(conn.getTenantId(), TRANSACTIONS_TABLE)
+      + " WHERE  (jsonb->>'fromFundId' = '"+ budget.getFundId() +"' OR jsonb->>'toFundId' = '"+ budget.getFundId() + "')"
+      + " AND (jsonb->>'fiscalYearId')::text = '" + budget.getFiscalYearId() + "' AND (jsonb->>'transactionType')::text = 'Allocation'";
+
+    return conn.execute(sql)
+      .onSuccess(ecList -> logger.info("deleteAllocationTransactions:: Allocation transaction for budget with id {} successfully deleted",
+        budget.getId()))
+      .onFailure(e -> logger.error("deleteAllocationTransactions:: Allocation transaction deletion by query {} failed for budget with id {}",
+        sql, budget.getId(), e))
+      .mapEmpty();
+  }
+
+  private Future<Void> checkTransactions(Budget budget, DBConn conn) {
+    logger.debug("checkTransactions:: Checking transactions with fund id {}", budget.getFundId());
+
+    String sql ="SELECT jsonb FROM "+ getFullTableName(conn.getTenantId(), TRANSACTIONS_TABLE)
+      + " WHERE  (jsonb->>'fromFundId' = '"+ budget.getFundId() + "' AND jsonb->>'fiscalYearId' = '" + budget.getFiscalYearId()
+      + "' OR jsonb->>'toFundId' = '"+ budget.getFundId() + "'"
+      + " AND jsonb->>'fiscalYearId' = '" + budget.getFiscalYearId() + "') AND ((jsonb->>'transactionType')::text<>'Allocation'"
```

Comment:
the same here about foreign keys, regarding transactionType - I see that gin index is created for this field, but is is not used now.
To make it in use we need to see declaration of this index and use exactly the same query that used for index creating.
In our case: lower(f_unaccent(jsonb ->> 'transactionType'::text))
Link for reference where we use function to use indexes: [Link](https://github.com/folio-org/mod-finance-storage/blob/master/src/main/resources/templates/db_scripts/budget_encumbrances_rollover.sql#L700) 
Of course we can do this improvement in scope of subsequent PR, as you prefer

Reply:
Interesting. I will try that as well.

Reply:
It would be more convenient to have an index without the lower(f_unaccent()) though, maybe something to do in another ticket...